src/main/java/net/fabricmc/installer/UnattendedMain.java

```diff
@@ -0,0 +1,182 @@
+import java.lang.invoke.MethodType;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+public final class UnattendedMain {
+    private static final String INSTALL_CONFIG_NAME = "install.properties";
+    private static final Path SERVER_DIR = Paths.get("", ".fabric", "server");
+
+    public static void main(String[] args) throws Throwable {
+        LaunchData launchData;
+        try {
+            launchData = initialise();
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to setup fabric server", e);
+        }
+
+        Objects.requireNonNull(launchData, "launchData is null, cannot proceed");
+
+        // Set the game jar path to bypass loader's own lookup
+        System.setProperty("fabric.gameJarPath", launchData.serverJar.toAbsolutePath().toString());
+
+        URLClassLoader launchClassLoader = new URLClassLoader(new URL[]{ launchData.launchJar.toUri().toURL() });
+
+        // Use method handle to keep the stacktrace clean
+        MethodHandle handle = MethodHandles.publicLookup().findStatic(launchClassLoader.loadClass(launchData.mainClass), "main", MethodType.methodType(void.class, String[].class));
+        handle.invokeExact(args);
+    }
+
+    // Validates and downloads/installs the server if required
+    private static LaunchData initialise() throws IOException {
+        Properties properties = readProperties();
+
+        String loaderVersion = Objects.requireNonNull(properties.getProperty("fabric-loader-version"), "no loader-version specified in " + INSTALL_CONFIG_NAME);
+        String gameVersion = Objects.requireNonNull(properties.getProperty("game-version"), "no game-version specified in " + INSTALL_CONFIG_NAME);
+
+        // 0.12 or higher is required
+        validateLoaderVersion(loaderVersion);
+
+        // Vanilla server jar
+        Path serverJar = SERVER_DIR.resolve(String.format("%s-server.jar", gameVersion));
+        // Includes the mc version as this jar contains intermediary
+        Path serverLaunchJar = SERVER_DIR.resolve(String.format("fabric-loader-server-%s-minecraft-%s", loaderVersion, gameVersion));
+
+        String mainClass = null;
+
+        if (Files.exists(serverJar) && Files.exists(serverLaunchJar)) {
+            try {
+                mainClass = readMainClass(serverLaunchJar);
+            } catch (IOException e) {
+                // Wont throw here, will try to reinstall
+                System.err.println("Failed to read main class from server launch jar: " + e.getMessage());
+            }
+        }
+
+        if (mainClass != null) {
+            // All seems good, no need to reinstall
+            return new LaunchData(serverJar, serverLaunchJar, mainClass);
+        }
```

Comment:
I'd move the above block into the `try {` above and remove the mainClass var. The catch block should also handle ZipError.